04-onion-routing.md

```diff
@@ -229,38 +239,97 @@ leaking its position in the route.
+    - MUST return an error if `encrypted_recipient_data` does not decrypt using the
+      blinding point as described in [Route Blinding](#route-blinding).
+    - If `payment_constraints` is present:
+      - MUST return an error if:
+        - the expiry is greater than `encrypted_recipient_data.payment_constraints.max_cltv_expiry`.
+        - the amount is below `encrypted_recipient_data.payment_constraints.htlc_minimum_msat`.
+    - If `allowed_features` is missing:
+      - MUST process the message as if it were present and contained an empty array.
+    - If `allowed_features` is present:
+      - MUST return an error if:
+        - `encrypted_recipient_data.allowed_features.features` contains an unknown feature bit (even if it is odd).
+        - the payment uses a feature not included in `encrypted_recipient_data.allowed_features.features`.
+    - If it is not the final node:
+      - MUST return an error if the payload contains other tlv fields than `encrypted_recipient_data` and `current_blinding_point`.
+      - MUST return an error if `encrypted_recipient_data` does not contain either `short_channel_id` or `next_node_id`.
+      - MUST return an error if `encrypted_recipient_data` does not contain `payment_relay`.
+      - MUST use values from `encrypted_recipient_data.payment_relay` to calculate `amt_to_forward` and `outgoing_cltv_value` as follows:
+        - `amt_to_forward = ((amount_msat - fee_base_msat) * 1000000 + 1000000 + fee_proportional_millionths - 1) / (1000000 + fee_proportional_millionths)`
+        - `outgoing_cltv_value = cltv_expiry - payment_relay.cltv_expiry_delta`
+    - If it is the final node:
+      - MUST return an error if the payload contains other tlv fields than `encrypted_recipient_data`, `current_blinding_point`, `amt_to_forward`, `outgoing_cltv_value` and `total_amount_msat`.
+      - MUST return an error if `amt_to_forward` or `outgoing_cltv_value` are not present.
+      - MUST return an error if `amt_to_forward` is below what it expects for the payment.
+  - Otherwise (it is not part of a blinded route):
+    - MUST return an error if `amt_to_forward` or `outgoing_cltv_value` are not present.
+    - if it is not the final node:
+      - MUST return an error if:
+        - `short_channel_id` is not present,
+        - it cannot forward the HTLC to the peer indicated by the channel `short_channel_id`.
+        - incoming `amount_msat` - `fee` < `amt_to_forward` (where `fee` is the advertised fee as described in [BOLT #7](07-routing-gossip.md#htlc-fees))
+        - `cltv_expiry` - `cltv_expiry_delta` < `outgoing_cltv_value`
+  - If it is the final node:
+    - MUST treat `total_msat` as if it were equal to `amt_to_forward` if it is not present.
     - MUST return an error if:
-       - incoming `amount_msat` != `amt_to_forward`.
-       - incoming `cltv_expiry` != `cltv_expiry_delta`.
+      - incoming `amount_msat` != `amt_to_forward`.
+      - incoming `cltv_expiry` != `cltv_expiry_delta`.
```

Comment:
Good catch, this is indeed a general issue that should be fixed directly on `master` asap, so I created #1032 to do that.

I think we need to keep `outgoing_cltv_value` in the onion payload for the final recipient, otherwise there would still a probing attack vector: an intermediate node could simply decrease the expiry of the htlc it sends. Intermediate nodes would reject this, while final nodes would accept it if it still complied with their `min_final_cltv_expiry_delta`.